tokio/src/io/util/read_line.rs

```diff
@@ -66,7 +65,14 @@ impl<R: AsyncBufRead + ?Sized + Unpin> Future for ReadLine<'_, R> {
             bytes,
             read,
         } = &mut *self;
-        read_line_internal(Pin::new(reader), cx, buf, bytes, read)
+        let ret = ready!(read_line_internal(Pin::new(reader), cx, buf, bytes, read));
+        if ret.is_err() {
+            // If an error occurred, put the data back, but only if it is valid utf-8.
+            if let Ok(string) = String::from_utf8(mem::replace(bytes, Vec::new())) {
+                **buf = string;
+            }
+        }
```

Comment:
Why is this processing outside of `read_line_internal`? ([`lines` also use `read_line_internal`](https://github.com/tokio-rs/tokio/blob/f48065910eca238d7ed644807f85c921deb12ec3/tokio/src/io/util/lines.rs#L93))

Reply:
I guess that's an artifact of being inspired by #2384. That said, we should probably consider if #2544 has any effects on this PR before moving forward.